staging/src/k8s.io/csi-api/pkg/crd/crd.go

```diff
@@ -14,106 +14,9 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// Package crd is only for running tests to verify the manifest files
+// in this package and the addons are in sync.
+// The manifest file is currently manually generated, in the future, we
+// should invest in tooling that will automatically generate the CRD
+// manifest from the CR schema.
 package crd
-
-import (
-	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	csiapiv1alpha1 "k8s.io/csi-api/pkg/apis/csi/v1alpha1"
-	"reflect"
-)
-
-// NOTE: the CRD functions here and the associated unit tests are non-ideal temporary measures in
-// release 1.12 in order to aid manual CRD installation. This installation will be automated in
-// subsequent releases and as a result this package will be removed.
-
-// CSIDriverCRD returns the CustomResourceDefinition for CSIDriver object.
-func CSIDriverCRD() *apiextensionsv1beta1.CustomResourceDefinition {
-	return &apiextensionsv1beta1.CustomResourceDefinition{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: csiapiv1alpha1.CsiDriverResourcePlural + "." + csiapiv1alpha1.GroupName,
-		},
-		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
-			Group:   csiapiv1alpha1.GroupName,
-			Version: csiapiv1alpha1.SchemeGroupVersion.Version,
-			Scope:   apiextensionsv1beta1.ClusterScoped,
-			Validation: &apiextensionsv1beta1.CustomResourceValidation{
-				OpenAPIV3Schema: &apiextensionsv1beta1.JSONSchemaProps{
-					Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-						"spec": {
-							Description: "Specification of the CSI Driver.",
-							Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-								"attachRequired": {
-									Description: "Indicates this CSI volume driver requires an attach operation," +
-										" and that Kubernetes should call attach and wait for any attach operation to" +
-										" complete before proceeding to mount.",
-									Type: "boolean",
-								},
-								"podInfoOnMountVersion": {
-									Description: "Indicates this CSI volume driver requires additional pod" +
-										" information (like podName, podUID, etc.) during mount operations.",
-									Type: "string",
-								},
-							},
-						},
-					},
-				},
-			},
-			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
-				Plural: csiapiv1alpha1.CsiDriverResourcePlural,
-				Kind:   reflect.TypeOf(csiapiv1alpha1.CSIDriver{}).Name(),
-			},
-		},
-	}
-}
-
-// CSINodeInfoCRD returns the CustomResourceDefinition for CSINodeInfo object.
-func CSINodeInfoCRD() *apiextensionsv1beta1.CustomResourceDefinition {
```

Comment:
its unclear to me why we are removing this, previously the `csinodeinfo.yaml` was being generated from this definition when you ran the `crd_test.go`. That was useful for validation.

Reply:
The code generated version was previously being used as the source of truth, and the YAML was generated from it. Two problems with that was 1) source of truth should be a YAML file not generated code per API machinery folks. 2) the generated yaml contained invalid fields (e.g. status), that resulted not being able to use the YAML in `kubectl create -f` without disabling validation.

So instead I'm moving the source of truth to YAML in `crd/manifests`. If any one wants a programmatic version of it, they can load it from the YAML file (see the integration test that does this).

Reply:
The source of truth should be "go IDL", i.e. types.go style struct definitions. yaml is also OK, although no tools will work on it. The code getting removed here was not an acceptable format for being a source of truth, so it's good that it's getting deleted.